lib/rules/indent.js

```diff
@@ -807,6 +892,96 @@ module.exports = {
             });
         }
 
+        /**
+        * Set the current unknown node if not set
+        * @param {ASTNode} node Node
+        * @returns {void}
+        */
+        function setCurrentUnknownNode(node) {
+            if (KNOWN_NODES.has(node.type)) {
+                return;
+            }
+
+            if (currentUnknownNode === null) {
+                currentUnknownNode = node;
+            }
+        }
+
+        /**
+        * On exit of unknown node append it to list of outermost unknown nodes
+        * @param {ASTNode} node Node
+        * @returns {void}
+        */
+        function saveOutermostUnknownNode(node) {
+            if (KNOWN_NODES.has(node.type)) {
+                return;
+            }
+
+            if (currentUnknownNode === node) {
+                unknownNodes.push(node);
+                currentUnknownNode = null;
+            }
+        }
+
+        /**
+        * Ignore first token of each line of within an unknown node if its
+        * offset does not depend on another token's offset. If the first token
+        * of a line's offset depends on the offset of another token, make the
+        * last dependent offset depend on the first token of its line.
+        * @param {ASTNode} node Node
+        * @returns {void}
+        */
+        function ignoreUnknownNode(node) {
+            const startLine = node.loc.start.line;
+            const endLine = node.loc.end.line;
+
+            for (let x = startLine; x <= endLine; x++) {
+                const firstTokenOfLine = tokenInfo.firstTokensByLineNumber.get(x);
+
+                if (!firstTokenOfLine) {
+                    continue;
+                }
+
+                const tokenInNodeRange = (node.range[0] <= firstTokenOfLine.range[0]) &&
+                    (node.range[1] >= firstTokenOfLine.range[1]);
+
+                if (!tokenInNodeRange) {
+                    continue;
+                }
+
+
+                const dependency = offsets.getFirstDependency(firstTokenOfLine);
+
+                if (!dependency) {
+                    offsets.ignoreToken(firstTokenOfLine);
+                } else {
+                    let currDependency = dependency,
+                        lastDependency = dependency;
+
+                    while ((currDependency = offsets.getFirstDependency(currDependency))) {
```

Comment:
Why is it necessary to iterate through dependencies here? It seems like this will ignore a token that has no dependencies at the top level, but that might not be desirable. For example, a custom node might be nested:

```js
function foo() {
  function bar() {
    abstract class X {
      public baz() {
        qux();
      }
    }
  }
}
```

It looks like this would end up ignoring the `function` token at the start of the program, but I think that would be incorrect behavior.

If I'm understanding the logic correctly, it seems like it would be better to stop this `while` loop at the last token which is still within the bounds of the unknown node.

But in that case, is the loop even necessary? It seems like it would be better to ignore the first token of every line within the node, if the token's first dependency when exiting the node is outside the node.